### DISCUSSION TITLE
Multi port IR Transmitter using Tasmota+ESP32

### DISCUSSION THREAD

**Discussion by nccchirag** [arendst/Tasmota#16100] [Q&A] [answered]
I'm very new to the tasmota world and I seek guidance for the question mentioned below.

Hardware Used
ESP32 pico dev board kit

**Module - IR Remote (1056	IRsend	IR interface)**
**Q** - IRSend component currently can be defined/assigned to one GPIO pin, how can I assign IRSend for multiple pins and transmit  IR command to specific assigned GPIO via console or via MQTT?

Inspiration product - _I intend to build a similar product using ESP32 + tasmota_
[Global Cache WiFi2IR](https://www.globalcache.com/files/docs/datasheet_itach_wifi.pdf) 
![image](https://user-images.githubusercontent.com/7569367/181273725-734986e8-6803-4fef-aa6b-372c413c4d11.png)

As shown above each port is addressed in the IR command API eg sendir, 1:**3** where 3 is the third port where the IR command would be fired

Can something similar be achieved using Tasmota readily?

Kindly guide
Thank you

**Comment by s-hadinger** [comment, marked as answer]
Not supported in Tasmota. It would be impossible on ESP8266, and would be doable on ESP32. However the underlying lib IRRemoteESP8266 has not been ported to use hardware RMT.

See https://github.com/crankyoldgit/IRremoteESP8266/discussions/1847

**Comment by barbudor** [comment]
As of today, tasmota supports only 1 single GPIO for IR
It doesn't prevent to drive an array of diodes to achieve 360° beam but the same signal will be sent in every directions.

I don't really see the need for different IR channels as the goal is actually to broadcaster to all receiveing devices in the same room.
For multiple rooms, multiply the the ESPs

**Comment by lalo-uy** [comment]
Maybe different carrier freq?

**Comment by s-hadinger** [comment]
Not even. The carrier frequency can change with each message

**Comment by sfromis** [comment]
What benefit would you see from separate IR transmit channels?

**Comment by nccchirag** [comment]
Thank you everyone for the responses!

More than hardware RMT it is about s/w implementation which does not define additional GPIO assignments eg IRSend1, IRSend2 etc.  for example, we can define a maximum of 8 relays / switches or buttons ... Similarly, it would be great to have IRSend1 ... IRSend4 Max 4 channels in the list of default components ... 

maybe I'm wrong in understanding how tasmota works... but if I assign IRSend to multiple pins then I think multiple instances are initialized at code level and each instance would receive FUNC_COMMAND event thereby the command would be transmitted on all pins initialized ... Please guide

Also, below are the points which can be possible benefits of using separate IR channels

1. **cost saving** -- as using ESP32 is an overkill for being used as a single channel IR blaster (360°)
2. blaster requires a **clear line of site** between the device and the IR blaster ... with separate channels one can use an Infrared patch which can be wired and hence providing robust operation without the requirement of "clear line of site"
3. most homes in India have similar brand devices eg living-dining would have **same make-model** AC and if we use one blaster, we cannot control them individually as both would be operated using the same IR command
4. with one blaster / transmitter per device increases the total number of Wi-Fi devices in one home adding up to the **maintenance** of each IP based device -- incl provisioning

Eg use case diagram
![image](https://user-images.githubusercontent.com/7569367/181432595-53419fc2-1c20-4e4c-9060-347dc05f7947.png)

Thank you anyway for all the assistance. I can safely assume this feature request would require custom development at code level and hopefully if I'm able to make the changes successfully then I shall give a PR. Please do share your views on this & whether it makes sense to modify the code to achieve the feature?

Lastly, I would like to take the opportunity to thank everyone in the Tasmota community & the collaborators/makers of Tasmota. It is one of the best architectured embedded open-source software ever! 🙏

**Reply by s-hadinger** [reply]
Let me clarify why I mentioned RMT. The current implementation is software only based on interrupts and tight timing. This means that you can't receive/send different streams in parallel (they would compete for the same CPU cycles). RMT would allow for concurrent sends/receives.

When it comes to sending IR, it could be possible to use different GPIOs as long as they are not used at the same time. You could indeed sequentially send on different GPIOs. Let me check if that's easy to add.

**Reply by nccchirag** [reply]
got it :) 
Yes, I need to transmit sequentially to different GPIO. Waiting for your reply 🙌
thank you

**Reply by barbudor** [reply]
The goal of IOT and ESP based devices is to get rid of wires by using dead simple function everywhere it is needed

1. Cost saving: barely valid argument. An ESP32c3 is very cheap and can be easily multiplied everywhere it is needed

2 & 3. If you don't want to/can't use 360°, stick 1 esp+ir directly on each apparatus to control. Don't wire.

4. How is it a problem? You can have hundreds of Tasmotized devices. Not a problem.

So from my personal point of view, it just doesn't make sense.

**Reply by nccchirag** [reply]
I appreciate your viewpoint however; I have a very different application scenario and hence the requirement. 
For eg - Sticking 1 esp at every device isn't practical as we require AC-DC power socket at each control device location (aesthetically & commercially) ... most Air-conditioners here in India are split type which do not have any 230V/power socket near the indoor unit as it is usually powered via the outdoor unit directly ... so, running a passive cable (just the signal patch) is easier

Wi-Fi signal strength is another area of concern as most Audio-Video devices are hidden in a cabinet and have very low to no wifi range

All your arguments are valid when it is DIY - do it yourself scenario (personal use) where-in the whole thing is installed and managed by one person. However, the moment we look at a commercial solution we need to optimize costs and maintenance across the board. 

I hope I'm able to explain it with somewhat clarity as to why I require the multi-channel IR emitter

thank you so much for the response though :)

**Reply by barbudor** [reply]
One little issue if someone implement an indexed IR GPIO is that you wouldn't be able to use 'IrSend1', 'IrSend2'.. As this is already in use in the IrSend command as a repetition argument (see the command syntax) so you would need an additional argument in the JSON (like `..., "output": 2,...`)
Not a big deal though

Note that you could already implement some kind of multichannel by using a single output and a demultiplexer such as 74HC238 driven by some pseudo Relay outputs.

**Reply by s-hadinger** [reply]
Yes I already spotted that index is already used. Although I don't like to do it, we can have a convention saying that 2xx is GPIO2, 3xx is GPIO3... I will also add a json attribute if the payload is JSON

**Reply by nccchirag** [reply]
can we have the index as 1:<ch>? JSON attribute can be channel: "1"
where ch = channel number or array index mapped to a GPIO pin

However, both the initialization and IR xdrv would require code changes? 
IRsend *irsend[4];
irsend[0] = new IRsend( ...

this is crude way but would something like the above work? Please guide

**Reply by s-hadinger** [reply]
Actually revisiting the code, we don't need to keep an array of `IRsend` objects. We can just reinstantiate a local object each time we send a message. This is the way it works for IRac. The constructor of `IRsend` doesn't do much.

This way we don't have any limit on the number of IR GPIOs, and it makes things simpler to manage in the code.

**Comment by s-hadinger** [comment]
This is now included in https://github.com/arendst/Tasmota/pull/16138

You can add `"Channel":1` or any value to send on different GPIOs.

Can you please test and report?

**Reply by sfromis** [reply]
While using `Channel` in the JSON format is a nice way of doing it, what about cases where `IRSend` needs to be raw format?

**Reply by s-hadinger** [reply]
As an advanced feature, I would only support it for json. I think that we may need to evolve json to also support raw format

**Reply by nccchirag** [reply]
thank you @s-hadinger for the lightning support. Will check and update soon!
You rock! 🙏

**Reply by nccchirag** [reply]
> This is now included in #16138
> 
> You can add `"Channel":1` or any value to send on different GPIOs.
> 
> Can you please test and report?

It works flawlessly! Thank you so much 🙏

**Reply by s-hadinger** [reply]
Good to know, thanks

**Reply by nccchirag** [reply]
> While using `Channel` in the JSON format is a nice way of doing it, what about cases where `IRSend` needs to be raw format?

I have modified the code as shown below to support channel in **GC** raw commands ... Similarly, we can do it for **RAW** commands too! May not be the best way but seems to work :)

**Format**
// IRsend gc#,xxx,xxx,...
// IRsend raw#,<freq>

where # is the channel number

NOTE: it won't work in case of standard raw command format // IRsend <freq>,<rawdata>,<rawdata> ...

![code](https://user-images.githubusercontent.com/7569367/182812615-09793234-7b89-45ca-bdb1-d5de1950b8f8.png)

**Reply by nccchirag** [reply]
@s-hadinger 
May I request you to guide regarding the below question ?
 
I need to access and control other peripherals from within the IR driver (xdrv). For eg simple use case is to turn ON an LED while the IR is being fired then turn it OFF when its done! Can I use digitalWrite(... or should I execute command by calling the function ExecuteWebCommand("LedPower2 1", SRC_WEBCONSOLE);

May be rules can be used here but I have compiled binary without support of rules & expression. Also, I wanted to learn to access low level hardware programmatically within the Tasmota architecture / framework.

Please guide

Thank you

**Reply by nccchirag** [reply]
I think, I have figured it out from sonoff xdrv code ... @s-hadinger you may ignore the above query ... Below is what worked for me!

ExecuteCommand("LedPower3 0", **SRC_REMOTE**); 

**xdrv_22_sonoff_ifan.ino**
        snprintf_P(svalue, sizeof(svalue), PSTR(D_CMND_FANSPEED " %d"), action);
        ExecuteCommand(svalue, SRC_REMOTE);

Thank you

**Comment by s-hadinger** [comment]
As mentioned, I will instead include raw/gc/bitstream encoding as JSON to make it the preferred format.

**Comment by nccchirag** [comment]
How can I add a new custom Command / New Settings Flag which can be toggled by SetOptionsXXX 1/0? 
Can anyone please guide me?

**Intent**
Enable & Disable IR Receive 
SetOption145	1 = Enable IR Receiver 0 = Disable IR Receiver

Even though SetOption38 is set to minimum, the receiver seems to capture some signal.

Thank you

**Reply by sfromis** [reply]
Ir you do not want IR receive, you need no new option for that, you can just unconfigure the `IRRecv` pin. Sending still works.

**Reply by nccchirag** [reply]
that is correct but, I wish to keep the configuration / template intact while adding enable/disable functionality via SetOptions
